server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransactionLog.java

```diff
@@ -0,0 +1,286 @@
+    if (this.log == origLog) {
+      this.log = newLog;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Get a string that provides a list of the transactions.
+   *
+   * @return a log dump
+   */
+  public String dumpLog() {
+    return this.log.dumpLog(extent);
+  }
+
+  @Override
+  public String toString() {
+    return dumpLog();
+  }
+
+  /**
+   * A transaction log consists of the original file set and its timestamp, a set of transactions,
+   * and the final set of files after applying the transations. This class is immutable.
+   */
+  private static class TransactionLog {
+    // The time stamp of the initial file set
+    private final long initialTs;
+    // the initial file set
+    private final StoredTabletFile[] initialFiles;
+    // the transactions
+    private final DatafileTransaction[] tabletLog;
+    // the final file set derived be applying the transactions to the initial file set
+    private final StoredTabletFile[] finalFiles;
+
+    public TransactionLog(Set<StoredTabletFile> files) {
+      this(files.toArray(new StoredTabletFile[0]));
+    }
+
+    private TransactionLog(StoredTabletFile[] files) {
+      this(System.currentTimeMillis(), files, new DatafileTransaction[0], files);
+    }
+
+    private TransactionLog(long initialTs, StoredTabletFile[] initialFiles,
+        DatafileTransaction[] tabletLog, StoredTabletFile[] finalFiles) {
+      this.initialTs = initialTs;
+      this.initialFiles = initialFiles;
+      this.tabletLog = tabletLog;
+      this.finalFiles = finalFiles;
+    }
+
+    /**
+     * This constructor will be the passed in log, adding the passed in transaction, and trimming
+     * the log to maxSize if needed.
+     *
+     * @param log The starting log
+     * @param transaction The new transaction
+     * @param maxSize The max transaction log size
+     */
+    public TransactionLog(TransactionLog log, DatafileTransaction transaction, int maxSize) {
+      // if the starting log is smaller than maxSize, then simply add the transaction to the end
+      if (log.tabletLog.length < maxSize) {
+        this.initialTs = log.initialTs;
+        this.initialFiles = log.initialFiles;
+        this.tabletLog = Arrays.copyOf(log.tabletLog, log.tabletLog.length + 1);
+        this.tabletLog[log.tabletLog.length] = transaction;
+        this.finalFiles = applyTransaction(log.finalFiles, transaction);
+      } else if (maxSize <= 0) {
+        // if the max size is 0, then return a log of 0 size, applying the transaction to the file
+        // set
+        this.initialTs = transaction.ts;
+        this.initialFiles = this.finalFiles = applyTransaction(log.finalFiles, transaction);
+        this.tabletLog = new DatafileTransaction[0];
+      } else {
+        // otherwise we are over the max size limit. Trim the transaction log and apply transactions
+        // appropriately to the initial and final file sets.
+        this.tabletLog = new DatafileTransaction[maxSize];
```

Comment:
Could possibly use this circular buffer for the transactions to avoid copying on each insert.

https://guava.dev/releases/15.0/api/docs/com/google/common/collect/EvictingQueue.html

Reply:
The documentation for the distruptor ring buffer is at

https://lmax-exchange.github.io/disruptor/user-guide/index.html

Reply:
I found that implementation not as safe as I wanted.  I created a simple implementation instead that gives me what I need.